### PR TITLE
fix `static mut` dealloc or freeze

### LINKED ISSUE #90: immutable static with raw pointer to mutable static makes mutable static immutable
>  a non-mutable static might contain a raw pointer into a static mut and this code would then freeze the static mut. Don't worry about it for now.

this code: https://github.com/solson/miri/pull/85#discussion_r88527276

Code example reproducing this issue:

```rust
#![allow(dead_code)]

static mut FOO: i32 = 42;
static BAR: Foo = Foo(unsafe { &FOO as *const _} );

struct Foo(*const i32);

unsafe impl Sync for Foo {}

fn main() {
    unsafe {
        assert_eq!(*BAR.0, 42);
        FOO = 5;
        assert_eq!(FOO, 5);
        assert_eq!(*BAR.0, 5);
    }
}
```

### LINKED ISSUE #118: `static mut`s deallocate their sub-allocations, making their pointers dangle
**Source**

```rust
static mut FOO: &'static i32 = &42i32;

pub fn main() {
    unsafe { *FOO; }
}
```

**Output and error**

```
TRACE:miri::step:  pushing stack frame for global: FOO
TRACE:miri::step:   _2 = const 42i32
TRACE:miri::eval_context:   _2: Undef
TRACE:miri::eval_context:   _2: Bytes(42)
TRACE:miri::step:   _1 = &_2
TRACE:miri::eval_context:   _1: Undef
TRACE:miri::eval_context:   _2: Bytes(42)
TRACE:miri::eval_context:   _1: Ptr(Pointer { alloc_id: AllocId(2), offset: 0 })
TRACE:miri::memory:   Alloc 2:    2a 00 00 00 (4 bytes)
TRACE:miri::step:   _0 = &(*_1)
TRACE:miri::eval_context:   _1: Ptr(Pointer { alloc_id: AllocId(2), offset: 0 })
TRACE:miri::memory:   Alloc 2:    2a 00 00 00 (4 bytes)
TRACE:miri::eval_context:   _1: Ptr(Pointer { alloc_id: AllocId(2), offset: 0 })
TRACE:miri::memory:   Alloc 2:    2a 00 00 00 (4 bytes)
TRACE:miri::lvalue:   deref to i32 on ByVal(Ptr(Pointer { alloc_id: AllocId(2), offset: 0 }))
TRACE:miri::step:   return
TRACE:miri::eval_context:  deallocating local
TRACE:miri::memory:  Alloc 2:    2a 00 00 00 (4 bytes)
DEBUG:miri::memory: deallocated : 2
TRACE:miri::step:  // bb0
TRACE:miri::step:  _2 = (*FOO)
TRACE:miri::eval_context:  _2: Undef
TRACE:miri::lvalue:  deref to i32 on ByVal(Ptr(Pointer { alloc_id: AllocId(2), offset: 0 }))
error: dangling pointer was dereferenced
 --> const-vec-of-fns.rs:4:14
  |
4 |     unsafe { *FOO; }
  |              ^^^^
```

### LINKED ISSUE #120: recursive static initialization is broken
I investigated the only "tried to modify constant memory" in #55 and it's due to the fact that Miri can't properly initialize statics that recursively refer to each other. This would be low priority since it only affects the unstable `#![feature(static_recursion)]`, but the feature looks like it will be [stable soon](https://github.com/rust-lang/rust/issues/29719).

E.g.

```rust
struct S(&'static S);
static S1: S = S(&S2);
static S2: S = S(&S1);
```

When you start initializing S1:

1. you'll start initializing S2 because it's required (`ConstantExtractor`)
2. write the final result for S2
3. freeze S2 and all allocations it references (this freezes S1)
4. pick up where you left off initializing S1
5. write the final result for S1 (fails because it's frozen)

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #90
fixes #118
fixes #120

### CODE REVIEW

File: src/error.rs [hunk 1]
```diff
@@ -118,9 +118,9 @@ impl<'tcx> Error for EvalError<'tcx> {
                 "cannot evaluate inline assembly",
             EvalError::TypeNotPrimitive(_) =>
                 "expected primitive type, got nonprimitive",
-            EvalError::ReallocatedFrozenMemory =>
+            EvalError::ReallocatedStaticMemory =>
                 "tried to reallocate frozen memory",
```
Comment (bjorn3) [inline src/error.rs]: "static memory"?

File: src/memory.rs [hunk 2]
```diff
@@ -38,9 +38,19 @@ pub struct Allocation {
     /// The alignment of the allocation to detect unaligned reads.
     pub align: u64,
     /// Whether the allocation may be modified.
-    /// Use the `freeze` method of `Memory` to ensure that an error occurs, if the memory of this
+    /// Use the `mark_static` method of `Memory` to ensure that an error occurs, if the memory of this
     /// allocation is modified in the future.
```
Comment (bjorn3) [inline src/memory.rs]: "(modified) or deallocated"?
Statics are no longer immutable.

File: src/memory.rs [hunk 3]
```diff
@@ -514,7 +527,11 @@ impl<'a, 'tcx> Memory<'a, 'tcx> {
                 }
             }
 
-            let immutable = if alloc.immutable { " (immutable)" } else { "" };
+            let immutable = match alloc.static_kind {
+                StaticKind::Mutable => "(static mut)",
+                StaticKind::Immutable => "(immutable)",
```
Comment (bjorn3) [inline src/memory.rs]: "(static)"

Comment (oli-obk) [inline src/memory.rs]: I like `immutable`, it could be a vtable or static or constant or promoted.

Comment (bjorn3) [inline src/memory.rs]: maybe use `global mutable` and `global immutable`

Comment (oli-obk) [inline src/memory.rs]: too long imo. Also, it's just the trace debug messages ;)

Comment (solson) [inline src/memory.rs]: These strings should start with spaces like the original.

**Pull request merged**